### PR TITLE
Implement permutation decoding

### PR DESCRIPTION
**Pull request by tirr-c**

### CODE REVIEW

File: jxl/src/headers/permutation.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        for idx in lehmer {
+            permutation.push(temp.remove(idx as usize));
+        }
+        permutation.extend(temp);
+
+        Ok(Self(permutation))
+    }
+}
+
+fn get_context(x: u32) -> usize {
```
Comment (veluca93) [inline jxl/src/headers/permutation.rs:77]: Would you mind moving this above the impl block?

File: jxl/src/headers/permutation.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        let mut temp = Vec::new();
+        temp.try_reserve((size - skip) as usize)?;
+        temp.extend((skip as usize)..(size as usize));
+
+        let mut permutation = Vec::new();
+        permutation.try_reserve(size as usize)?;
+
+        permutation.extend(0..skip as usize);
+        for idx in lehmer {
+            permutation.push(temp.remove(idx as usize));
```
Comment (veluca93) [inline jxl/src/headers/permutation.rs:69]: Isn't this quadratic?

File: jxl/src/headers/permutation.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        size: u32,
+        skip: u32,
+        end: u32,
+        mut read: impl FnMut(usize) -> Result<u32>,
+    ) -> Result<Self> {
+        if end > size - skip {
+            return Err(Error::InvalidPermutationSize { size, skip, end });
+        }
+
+        let mut lehmer = Vec::new();
```
Comment (veluca93) [inline jxl/src/headers/permutation.rs:43]: Might be worth moving the un-lehmering to another function

**Review by veluca93 — APPROVED** [review]

**Comment by mo271** [comment]
changes were integrated in #36

**Pull request closed without merging**